File: 686_Repeated_String_Match.py
_author_ = 'jake'
_project_ = 'leetcode'

# https://leetcode.com/problems/repeated-string-match/
# Given two strings A and B, find the minimum number of times A has to be repeated such that B is a substring of it.
# If no such solution, return -1.

# Repeat A until it is at least the length of B. If B is not a substring, then test with another copy of A added.
# More copies of A are not required because at least one copy of A will not be used.
# Time - O(n + m)
# Space - O(n + m)

class Solution(object):
    def repeatedStringMatch(self, A, B):
        """
        :type A: str
        :type B: str
        :rtype: int
        """
        if set(B) - set(A):         # early return
            return -1

        div, mod = divmod(len(B), len(A))

        if mod != 0:
            div += 1

        for i in range(2):
            if B in A * (div + i):
                return div + i

        return -1

File: java/686_Repeated_String_Match.java
/*
https://leetcode.com/problems/repeated-string-match/
Given two strings A and B, find the minimum number of times A has to be repeated such that B is a substring of it.
If no such solution, return -1.
For example, with A = "abcd" and B = "cdabcdab". Return 3, because by repeating A three times (“abcdabcdabcd”),
B is a substring of it; and B is not a substring of A repeated two times ("abcdabcd").

Concatenate n copies of A until it is at least the length of B. If B is a substing of any number of copies of A it must
be a substring of the n copies or n + 1 copies, since n + 2 copies means there must be a whole unused copy of A.
Time - O(n**2) where n is the length of B.
Space - O(n)
*/

class Solution {
    public int repeatedStringMatch(String A, String B) {

        int n = 1;
        StringBuilder sb = new StringBuilder(A);
        while (sb.length() < B.length()) {
            sb.append(A);
            ++n;
        }

        if (sb.indexOf(B) != -1)
            return n;

        if (sb.append(A).indexOf(B) != -1)
            return n + 1;

        return -1;
    }
}
